solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java

```diff
@@ -1017,19 +1023,21 @@ public String getBaseURL() {
   }
 
   private static class AsyncTracker {
-    private static final int MAX_OUTSTANDING_REQUESTS = 1000;
+    private static final int DEFAULT_MAX_OUTSTANDING_REQUESTS = 1000;
 
     // wait for async requests
     private final Phaser phaser;
+    private final int maxOutstandingRequests;
     // maximum outstanding requests left
     private final Semaphore available;
     private final Request.QueuedListener queuedListener;
     private final Response.CompleteListener completeListener;
 
-    AsyncTracker() {
+    AsyncTracker(int maxOutstandingRequests) {
       // TODO: what about shared instances?
+      this.maxOutstandingRequests = maxOutstandingRequests;
       phaser = new Phaser(1);
-      available = new Semaphore(MAX_OUTSTANDING_REQUESTS, false);
+      available = new Semaphore(maxOutstandingRequests, false);
```

Comment:
Not related to your review, but this semaphore is not working at all. `queuedListener` is called after the request may have been rejected, which essentially makes it useless.

Reply:
I'm all but certain that the semaphore _is_ working -- in fact I think it's the only thing that practically needed to be changed in our [local mitigation](https://github.com/cowpaths/fullstory-solr/pull/124). Notably, that mitigation changed `getMaxRequestsQueuedPerDestination()` to return a constant (3k -- effectively unchanged from current `main`) independent of the configuration of `maxOutstandingRequests`. There's some [relevant discussion](https://github.com/cowpaths/fullstory-solr/pull/124#discussion_r1294915329) there that explores the distinction between `maxOutstandingRequests` as enforced by the `available` Semaphore (across all destinations), vs. `getMaxRequestsQueuedPerDestination()` which, as the method name suggests, is per destination.

Reply:
I have already observed in production exceptions thrown because the internal queue of the HTTP2 client was full. It should not have been the case if the semaphore was working correctly, should it? When I was debugging the code, I found out that the `queuedListener` is called after the request had a change to get rejected, which means that the semaphore does not have a chance to protected this from happening. Or am I missing the point of this semaphore?

Reply:
It's been a while since I was in this code, but iirc (and thinking back to the "relevant discussion" comment linked above), there are _2_ queues: 1 is a per-destination buffer (before send), the other is the actual "request enqueued to be sent" queue, which is the one to which the semaphore applies, and covers _all_ destinations.

So I suspect that what happened was that the second queue (to which the semaphore applies) is actually throttling the request, which then causes the buffers to back up to the point where the requests get proactively rejected, which is what you would have been seeing in production.